schema-engine-tag/src/main/java/org/apache/iotdb/db/metadata/tagSchemaRegion/tagIndex/TagInvertedIndex.java

```diff
@@ -0,0 +1,193 @@
+import org.apache.iotdb.db.metadata.tagSchemaRegion.tagIndex.Request.InsertionRequest;
+import org.apache.iotdb.db.metadata.tagSchemaRegion.tagIndex.Request.QueryRequest;
+import org.apache.iotdb.db.metadata.tagSchemaRegion.tagIndex.memtable.MemTableGroup;
+import org.apache.iotdb.db.metadata.tagSchemaRegion.tagIndex.wal.WALEntry;
+import org.apache.iotdb.db.metadata.tagSchemaRegion.tagIndex.wal.WALManager;
+import org.apache.iotdb.lsm.engine.LSMEngine;
+import org.apache.iotdb.lsm.engine.LSMEngineBuilder;
+
+import org.roaringbitmap.RoaringBitmap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/** tag inverted index, tag is <tagkey,tagValue> and id is int32 auto increment id */
+public class TagInvertedIndex implements ITagInvertedIndex {
+
+  // This file records the wal log
+  private static final String WAL_FILE_NAME = "tag_inverted_index.log";
+
+  private static final Logger logger = LoggerFactory.getLogger(TagInvertedIndex.class);
+
+  // Manage configuration information of tag schema region
+  private static final TagSchemaConfig tagSchemaConfig =
+      TagSchemaDescriptor.getInstance().getTagSchemaConfig();
+
+  // Directly use the lsm engine that comes with the lsm framework to implement the tag inverted
+  // index
+  LSMEngine<MemTableGroup> lsmEngine;
+
+  /**
+   * initialization method
+   *
+   * @param schemaDirPath schema dirPath
+   */
+  public TagInvertedIndex(String schemaDirPath) {
+    try {
+      WALManager walManager =
+          new WALManager(
+              schemaDirPath,
+              WAL_FILE_NAME,
+              tagSchemaConfig.getWalBufferSize(),
+              new WALEntry(),
+              false);
+      // root memory node, used to manage working and immutableMemTables
+      MemTableGroup memTableGroup =
+          new MemTableGroup(tagSchemaConfig.getNumOfDeviceIdsInMemTable());
+
+      // build lsm engine
+      lsmEngine =
+          new LSMEngineBuilder<MemTableGroup>()
+              .buildLSMManagers("org.apache.iotdb.db.metadata.tagSchemaRegion.tagIndex", walManager)
+              .buildRootMemNode(memTableGroup)
+              .build();
+
+      // recover the lsm engine
+      lsmEngine.recover();
+    } catch (Exception e) {
+      logger.error(e.getMessage());
+    }
+  }
+
+  /**
+   * insert tags and device id
+   *
+   * @param tags tags like: <tagKey,tagValue>
+   * @param id INT32 device id
+   */
+  @Override
+  public synchronized void addTags(Map<String, String> tags, int id) {
+    try {
+      for (Map.Entry<String, String> tag : tags.entrySet()) {
+        InsertionRequest insertionRequest =
+            new InsertionRequest(generateKeys(tag.getKey(), tag.getValue()), id);
+        lsmEngine.insert(insertionRequest);
+      }
+    } catch (Exception e) {
+      logger.error(e.getMessage());
+    }
```

Comment:
There's no exception thrown in the interface definition. Please polish the exception handler mechanism. The exception shall be thrown or passed in execution result.